workspaces/mcp-chat/plugins/mcp-chat-backend/src/services/MCPClientServiceImpl.ts

```diff
@@ -0,0 +1,471 @@
+            },
+          });
+        }
+
+        // Connect the client with the appropriate transport
+        await client.connect(transport);
+
+        const toolsResult = await client.listTools();
+
+        const serverTools: ServerTool[] = toolsResult.tools.map(tool => ({
+          type: 'function',
+          function: {
+            name: tool.name,
+            description: tool.description ?? '', // Ensure string
+            parameters: tool.inputSchema,
+          },
+          serverId: serverConfig.name, // Track which server this tool belongs to
+        }));
+
+        allTools.push(...serverTools);
+        this.mcpClients.set(serverConfig.name, client);
+
+        // Record successful connection
+        serverResults.push({
+          ...baseServerConfig,
+          status: {
+            valid: isValid,
+            connected: true,
+          },
+        });
+
+        this.logger.info(
+          `MCP Server '${serverConfig.name}' connected via ${
+            serverConfig.type
+          } with tools: ${serverTools.map(t => t.function.name).join(', ')}`,
+        );
+      } catch (error) {
+        const errorMessage =
+          error instanceof Error ? error.message : String(error);
+
+        // Record failed connection
+        serverResults.push({
+          ...baseServerConfig,
+          status: {
+            valid: isValid,
+            connected: false,
+            error: errorMessage,
+          },
+        });
+
+        this.logger.warn(
+          `Failed to connect to MCP server '${serverConfig.name}': ${errorMessage}`,
+        );
+      }
+    }
+
+    this.tools = allTools;
+    this.connected = true;
+
+    const connectedServers = serverResults.filter(
+      s => s.status.connected,
+    ).length;
+    const totalServers = serverConfigs.length;
+    const failedServers = serverResults.filter(s => !s.status.connected);
+
+    if (failedServers.length > 0) {
+      this.logger.info(
+        `MCP initialization completed: ${connectedServers}/${totalServers} servers connected successfully. Failed servers: ${failedServers
+          .map(s => s.name)
+          .join(', ')}`,
+      );
+    } else {
+      this.logger.info(
+        `All MCP servers connected successfully. Total tools: ${this.tools.length}`,
+      );
+    }
+
+    return serverResults;
+  }
+
+  async processQuery(
+    messagesInput: any[],
+    enabledTools: string[] = [],
+  ): Promise<{
+    reply: string;
+    toolCalls: ToolCall[];
+    toolResponses: any[];
+  }> {
+    // Only add system message if one doesn't already exist
+    const messages: ChatMessage[] = [...messagesInput];
+    if (messages.length === 0 || messages[0].role !== 'system') {
+      messages.unshift({
+        role: 'system',
+        content:
+          "You are a helpful assistant. When using tools, provide a clear, readable summary of the results rather than showing raw data. Focus on answering the user's question with the information gathered.",
```

Comment:
Just an idea, would it make sense to be able to configure the system prompt using `Config`?

Reply:
Thanks, have implemented it.
You can now set this directly from the config.